impl/src/main/java/org/ehcache/internal/classes/ClassInstanceProvider.java

```diff
@@ -84,25 +84,44 @@ private T newInstance(K alias, ClassInstanceConfiguration<T> config) {
         return null;
       }
     }
-    try {
-      T instance = invokeConstructor(config.getClazz(), config.getArguments());
-      created.add(instance);
-      return instance;
-    } catch (InstantiationException e) {
-      throw new RuntimeException(e);
-    } catch (IllegalAccessException e) {
-      throw new RuntimeException(e);
-    } catch (NoSuchMethodException e) {
-      throw new RuntimeException(e);
-    } catch (InvocationTargetException e) {
-      throw new RuntimeException(e);
+    
+    T instance = null;
+    
+    if(config.getInstance() != null) {
+      instance = config.getInstance();
+    } else {
+      try {
+        instance = invokeConstructor(config.getClazz(), config.getArguments());
+      } catch (InstantiationException e) {
+        throw new RuntimeException(e);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      } catch (InvocationTargetException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    AtomicInteger currentCount = providedVsCount.get(instance);
+    if(currentCount == null) {
+      providedVsCount.put(instance, new AtomicInteger(1));
+    } else {
+      currentCount.incrementAndGet();
     }
+    return instance;
   }
 
   protected void releaseInstance(T instance) throws IOException {
-    if(!created.remove(instance)) {
+    if(providedVsCount.containsKey(instance)) {
```

Comment:
I believe it makes no sense to even try to do reference counting - for no added value as there is no conditional behaviour linked to it.
However, this discussion made me realise that we want the `created` type to be looking for instance equality not object equality, so you probably need to wrap a `IdentityHashMap` as a `Set` as your type there. And as your usage of an `AtomicInteger` indicates, the data structure must be thread safe and so needs even more wrapping to be synchronised - as this is good enough for lifecycle related methods.